### PR TITLE
Add typings for DataUrl

### PR DESCRIPTION
**Pull request by loichuder**

### CODE REVIEW

File: src/silx/io/url.py [hunk 1]
```diff
@@ -107,32 +134,33 @@ class DataUrl(object):
...
+        data_path: str | None = None,
+        data_slice: Tuple[SliceLike, ...] | None = None,
+        scheme: str | None = None,
+    ):
         self.__is_valid = False
         if path is not None:
-            assert(file_path is None)
-            assert(data_path is None)
-            assert(data_slice is None)
-            assert(scheme is None)
```
Comment (loichuder) [inline src/silx/io/url.py:135]: Some changes are due to auto-formatting by black. Sorry !

File: src/silx/io/url.py [hunk 2]
```diff
@@ -197,60 +228,41 @@ def __check_validity(self):
...
-                            return None
-                        else:
-                            return int(my_str)
-                    sss = string.split(':')
-                    start = get_value(sss[0])
-                    stop = get_value(sss[1] if len(sss) > 1 else None)
-                    step = get_value(sss[2] if len(sss) > 2 else None)
-                    return slice(start, stop, step)
-            else:
-                return int(string)
```
Comment (loichuder) [inline src/silx/io/url.py:231]: Moved out of the method and renamed `string_to_slice` for consistency with other functions that do the reverse operation.

File: src/silx/io/url.py [hunk 3]
```diff
@@ -197,60 +228,41 @@ def __check_validity(self):
...
         """Parse the path and initialize attributes.
 
-        :param str path: Path representing the URL.
+        :param path: Path representing the URL.
         """
         self.__path = path
         # only replace if ? not here already. Otherwise can mess sith
         # data_slice if == ::2 for example
-        if '?' not in path:
+        if "?" not in path:
```
Comment (loichuder) [inline src/silx/io/url.py:265]: Black black black

File: src/silx/io/url.py [hunk 4]
```diff
@@ -381,32 +387,21 @@ def is_absolute(self):
...
+    def file_path(self) -> str:
+        """Returns the path to the file containing the data."""
         return self.__file_path
 
-    def data_path(self):
-        """Returns the path inside the file to the data.
-
-        :rtype: str
-        """
+    def data_path(self) -> str | None:
```
Comment (loichuder) [inline src/silx/io/url.py:394]: `data_path` can be `None` if `path` is used to construct the `DataUrl`.

File: src/silx/io/url.py [hunk 5]
```diff
@@ -381,32 +387,21 @@ def is_absolute(self):
...
+        """Returns the path inside the file to the data."""
         return self.__data_path
 
-    def data_slice(self):
+    def data_slice(self) -> Tuple[SliceLike, ...] | None:
```
Comment (loichuder) [inline src/silx/io/url.py:398]: `data_slice` can also be `None` if no slice is specified.

File: src/silx/io/url.py [hunk 6]
```diff
@@ -381,32 +388,21 @@ def is_absolute(self):
...
-        """Returns the path inside the file to the data.
-
-        :rtype: str
-        """
+    def data_path(self) -> str | None:
+        """Returns the path inside the file to the data."""
         return self.__data_path
 
-    def data_slice(self):
+    def data_slice(self) -> tuple[SliceLike, ...] | None:
```
Comment (loichuder) [inline src/silx/io/url.py:399]: `data_slice` can also be `None` if no slice is specified.

**Review by t20100 — APPROVED** [review]
Thanks!

**Pull request merged**